Add GNAN model to PyG

codearky opened the pull request:
This PR adds the GNAN (https://arxiv.org/pdf/2406.01317) model and a dedicated dataloader to PyG

Commit:
Add GNAN model with tests, dataloader and mutagenicity example

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

puririshi98 committed:
Merge branch 'master' into master

Commit:
add feature groups

Commit:
add another example to grouping features

Commit:
add node importance to gnan model

Commit:
add tests for feature groups and node importance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
remove loading dataset from pickle for security reasins

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

puririshi98 committed:
Merge branch 'master' into master

Commit:
add test and fixed pre-commit errors

Commit:
add tests for gnan dataloader

Commit:
add to changelog

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

puririshi98 committed:
Merge branch 'master' into master

puririshi98 committed:
Merge branch 'master' into master

puririshi98 committed:
Merge branch 'master' into master

puririshi98 committed:
Merge branch 'master' into master

puririshi98 committed:
Merge branch 'master' into master

puririshi98 reviewed (changes requested):
LGTM at a high level, but please lets move the example to examples/explain.
Please also add a sentence to examples/README.md to mention, "For examples on explainability with GNNs, see examples/explain."
Then please also update examples/explain/README.md to have a short explanation of what the new example does. 
Once thats done I will do a deeper review and can merge

Commit:
edge index to int64

codearky committed:
Move `gnan_graph_mutagenicity.py` example script and update README files

codearky committed:
Merge branch 'master' of github.com:pyg-team/pytorch_geometric

codearky committed:
Rename GNAN model to MGNAN

The model was published as M-GNAN, an extension of GNAN to multivariate
shape functions. Renames the class TensorGNAN -> MGNAN, GNANDataLoader
-> MGNANDataLoader, GNANCollater -> MGNANCollater, and updates the
corresponding module / test / example file names accordingly.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

puririshi98 reviewed (changes requested):
LGTM, but can you share a log of the the e2e run of `examples/explain/mgnan_graph_mutagenicity.py` inside the nvidia PyG container to make sure it works as expected:  https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pyg/tags?version=26.03-py3

puririshi98 commented:
@codearky

codearky commented:
[mgnan-ngc-477482.full.log](https://github.com/user-attachments/files/28636754/mgnan-ngc-477482.full.log)
@puririshi98 Sorry for the delay, let me know if this satisfy the requirement

puririshi98 committed:
Merge branch 'master' into master

puririshi98 reviewed (approved):
LGTM

puririshi98 commented:
@akihironitta to merge

akihironitta commented on torch_geometric/loader/mgnan_dataloader.py:
We cannot place this new addition here since it's too specific to the use case. Since this dataloader is just a PyTorch dataloader with a simple custom `collate_fn`, we'd need to have this in the example script.

codearky replied:
Refactored to use a plain PyTorch DataLoader with a custom collater (MGNANCollater) that now lives in the example script.

akihironitta commented on torch_geometric/llm/models/g_retriever.py:
```diff
@@ -96,7 +96,7 @@ def encode(
     ) -> Tensor:
         x = x.to(self.llm.device)
-        edge_index = edge_index.to(self.llm.device)
+        edge_index = edge_index.to(torch.int64).to(self.llm.device)
```
If you have to cast `edge_index` here, we must be doing something wrong on the caller side. I'd suggest fixing it there instead of here.

codearky replied:
I removed this change as it has nothing to do with the pure GNAN case (it came from another layer of integration of GNAN with the G-Retriever pipeline - which is unnecessary in this PR and will be handled separately).

codearky committed:
Revert stray edge_index int64 cast in GRetriever

This change was orphaned from a different branch (GNAN + GRetriever
integration) and doesn't belong in the MGNAN PR. Per review feedback,
if edge_index dtype is wrong it should be fixed at the caller, not
worked around inside the model.

codearky committed:
Move MGNAN collate logic into the example, drop dedicated dataloader

Per review: MGNANDataLoader/MGNANCollater are too specific to the M-GNAN
training pipeline to live in torch_geometric.loader. Replace them with a
plain torch.utils.data.DataLoader and a module-level `mgnan_collate`
function inside examples/explain/mgnan_graph_mutagenicity.py that
block-diagonally stacks node_distances and normalization_matrix.

Removes torch_geometric/loader/mgnan_dataloader.py and the corresponding
test/loader/test_mgnan_loader.py; the example end-to-end exercises the
same code path.

codearky committed:
Restore MGNANCollater follow_batch/exclude_keys + hasattr guards

Keep the collater minimally diverged from the original
torch_geometric.loader.mgnan_dataloader.MGNANCollater: preserve the
follow_batch / exclude_keys constructor args and the per-attribute
hasattr guards. Only difference vs original: use torch.block_diag
instead of the hand-rolled _create_block_diagonal_matrix helper, since
it's bytewise identical for the square float matrices this pipeline
produces.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

codearky commented:
Following recent changes I rerun the e2e example
[mgnan-ngc-499052.full.log](https://github.com/user-attachments/files/28709656/mgnan-ngc-499052.full.log)

akihironitta committed:
Merge remote-tracking branch 'origin/master' into codearky/master

puririshi98 commented:
@akihironitta  to review

akihironitta committed:
update

akihironitta committed:
update

akihironitta committed:
update

akihironitta commented on examples/explain/mgnan_graph_mutagenicity.py:
```diff
@@ -0,0 +1,270 @@
+    indices = list(range(len(dataset)))
+    random.shuffle(indices)
+    n_train = int(0.8 * len(indices))
+    n_val = int(0.1 * len(indices))
```
```suggestion
    n_train = int(0.8 * len(dataset))
    n_val = int(0.1 * len(dataset))
```

akihironitta commented on examples/explain/mgnan_graph_mutagenicity.py:
```diff
@@ -0,0 +1,270 @@
+    n_train = int(0.8 * len(indices))
+    n_val = int(0.1 * len(indices))
+
+    # print("=" * 60)
```
```suggestion
```

akihironitta commented on torch_geometric/contrib/nn/models/mgnan.py:
```diff
@@ -0,0 +1,322 @@
+]
+
+
+def _init_weights(module: nn.Module, std: float = 1.0):
```
```suggestion
def _init_weights(module: nn.Module, std: float = 1.0) -> None:
```

akihironitta reviewed (approved):
I skimmed through the changes, LGTM!

akihironitta committed:
Apply suggestions from code review

akihironitta committed:
Merge branch 'master' into master
